### PR TITLE
script_tag: cosmetic for value-less attributes

### PR DESCRIPTION
**Pull request by xlii-chl**
**Description**

For `<script />` elements, some attributes are usually written without values, for example :
`<script defer async src="..." />`

This is purely cosmetic as boolean attributes should also be accepted with an empty
value or a value identical to the attribute's name :
https://dev.w3.org/html5/pf-summary/infrastructure.html#boolean-attribute

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by kenjis** [comment]
Thank you for sending a PR!

It seems you forgot GPG signing.
https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#signing

**Comment by kenjis** [comment]
@xlii-chl There are some coding style violations. Please fix them.
See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#php-style

```diff
   1) CodeIgniter4/system/Helpers/html_helper.php (is_null, unary_operator_spaces, not_operator_with_successor_space)
      ---------- begin diff ----------
--- /home/runner/work/CodeIgniter4/CodeIgniter4/system/Helpers/html_helper.php
+++ /home/runner/work/CodeIgniter4/CodeIgniter4/system/Helpers/html_helper.php
@@ -208,7 +208,7 @@
                 }
             } else {
                 // for attributes without values, like async or defer, use NULL.
-                $script .= $k . (is_null($v) ? ' ' : '="' . $v . '" ');
+                $script .= $k . (null === $v ? ' ' : '="' . $v . '" ');
             }
         }
 

      ----------- end diff -----------

   2) CodeIgniter4/tests/system/Helpers/HTMLHelperTest.php (trim_array_spaces)
      ---------- begin diff ----------
--- /home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Helpers/HTMLHelperTest.php
+++ /home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Helpers/HTMLHelperTest.php
@@ -250,7 +250,7 @@
 
     public function testScriptTagWithSrc()
     {
-        $target   = [ 'src' => 'http://site.com/js/mystyles.js' ];
+        $target   = ['src' => 'http://site.com/js/mystyles.js'];
         $expected = '<script src="http://site.com/js/mystyles.js" type="text/javascript"></script>';
         $this->assertSame($expected, script_tag($target));
     }
@@ -257,7 +257,7 @@
 
     public function testScriptTagWithSrcWithoutProtocol()
     {
-        $target   = [ 'src' => 'js/mystyles.js' ];
+        $target   = ['src' => 'js/mystyles.js'];
         $expected = '<script src="http://example.com/js/mystyles.js" type="text/javascript"></script>';
         $this->assertSame($expected, script_tag($target));
     }
@@ -264,7 +264,7 @@
 
     public function testScriptTagWithSrcAndAttributes()
     {
-        $target   = [ 'src' => 'js/mystyles.js', 'charset' => 'UTF-8', 'defer' => '', 'async' => null ];
+        $target   = ['src' => 'js/mystyles.js', 'charset' => 'UTF-8', 'defer' => '', 'async' => null];
         $expected = '<script src="http://example.com/js/mystyles.js" charset="UTF-8" defer="" async type="text/javascript"></script>';
         $this->assertSame($expected, script_tag($target));
     }

      ----------- end diff -----------
```
https://github.com/codeigniter4/CodeIgniter4/runs/5892697479?check_suite_focus=true

**Comment by xlii-chl** [comment]
Hello,

thanks for reviewing.

Sorry about the misses, I thought the gpg-signing etc. were for the core members or reviewers. Since signing modify the commit, I'll force-push in a bit if it's okay with you.

I'll put the `is_null($v) ? ...` → `null === $v ? ...` change in a separate commit (it seemed kinda less readable to me : needs to recall the operator precedence) so you can omit it if you feel the same way.

**Comment by kenjis** [comment]
Fore push is okay.

`git rebase` is recommended way to sync the PR branch.
See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#pushing-your-branch

> I'll put the is_null($v) ? ... → null === $v ? ... change in a separate commit (it seemed kinda less readable to me : needs to recall the operator precedence) so you can omit it if you feel the same way.

The coding style change is a must. If you don't fix it, GitHub Action check fails, and we can't merge this PR.

**Comment by kenjis** [comment]
@xlii-chl Thank you for updating!

One last thing, we need the User Guide update.
https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#documentation

**Comment by xlii-chl** [comment]
Hello,

For the doc, I've added a mention in the example:
```diff
--- a/user_guide_src/source/helpers/html_helper/011.php
+++ b/user_guide_src/source/helpers/html_helper/011.php
@@ -1,6 +1,6 @@
 <?php
 
-$script = ['src' => 'js/printer.js'];
+$script = ['src' => 'js/printer.js', 'defer' => null];
 
 echo script_tag($script);
-// <script src="http://site.com/js/printer.js" type="text/javascript"></script>
+// <script src="http://site.com/js/printer.js" defer type="text/javascript"></script>
```
For such a simple thing, it seemed enough to me but do you feel we should add more ? Since I am not a native english speaker, I may need help to write something good.

Or do you mean the Changelog ?
This change may be a little trivial for that, but what about, in the Enhancements section :
> `script_tag()` in the html helpers now able to concisely write boolean attributes : `<script src="..." defer />`.

**Comment by kenjis** [comment]
I thought we need explanation for the parameter `$src`, but as you say the sample code is enough.

Please update:
1) Parameters $src type
https://codeigniter4.github.io/CodeIgniter4/helpers/html_helper.html#script_tag

`mixed` is vague and useless.

```
    :param  array|string|null  $src: The source name of a JavaScript file
``` 

2) Change Log
https://codeigniter4.github.io/CodeIgniter4/changelogs/v4.2.0.html#enhancements

I think all enhancements should be included the changelog, otherwise users don't know the enhancements.

How about like this?
```
- The boolean attribute can now be written concisely by passing ``null`` to the ``script_tag()`` parameter ``$src`` in the HTML helper. See :php:func:`script_tag`'s sample code.
```

And could you also update the param type in PHPDoc? It is not `mixed` but `array|string|null`.
https://github.com/codeigniter4/CodeIgniter4/blob/a8d5b894c9ae17f1a8c8721eae44c0da22bec87c/system/Helpers/html_helper.php#L192
Don't forget to run `composer cs-fix`.

File: tests/system/Helpers/HTMLHelperTest.php [hunk 1]
```diff
@@ -248,6 +248,27 @@ public function testScriptTagWithIndexpage()
...
+    public function testScriptTagWithSrcWithoutProtocol()
+    {
+        $target   = ['src' => 'js/mystyles.js'];
+        $expected = '<script src="http://example.com/js/mystyles.js" type="text/javascript"></script>';
+        $this->assertSame($expected, script_tag($target));
+    }
+
+    public function testScriptTagWithSrcAndAttributes()
+    {
+        $target   = ['src' => 'js/mystyles.js', 'charset' => 'UTF-8', 'defer' => '', 'async' => null];
```
Comment (paulbalandan) [inline tests/system/Helpers/HTMLHelperTest.php:267]: is `'defer' => ''` intended? i assume this should be `null`?

Comment (kenjis) [reply tests/system/Helpers/HTMLHelperTest.php:267]: Oh, probably it is intended to test the behavior of the value `''`,
but `defer=""` is invalid.

Comment (xlii-chl) [reply tests/system/Helpers/HTMLHelperTest.php:267]: Hello,

yes, `'defer' => ''` is indeed intended to test the behaviour of the value `''`. But note that `defer=""` is a valid writing of the attribute:
* https://dev.w3.org/html5/spec-LC/common-microsyntaxes.html#boolean-attributes
* https://www.w3.org/TR/html401/intro/sgmltut.html#h-3.3.4.2

Given my understanding, all the following writings are valid and equivalent:
```
<script src="..." defer></script>
<script src="..." defer=""></script>
<script src="..." defer="defer"></script>
```

Doc updated using the W3 "minimized form" terminology, and I added a test on calling `script_tag()` without any argument.

Comment (kenjis) [reply tests/system/Helpers/HTMLHelperTest.php:267]: https://html.spec.whatwg.org/#boolean-attributes
Yes, @xlii-chl is correct.

**Review by paulbalandan — APPROVED** [review]

**Review by kenjis — APPROVED** [review]
LGTM

**Pull request merged**